### PR TITLE
Add a layer of abstraction to committing

### PR DESCRIPTION
**Pull request by jdavcs**
Ref: [SQLAlchemy documentation](https://docs.sqlalchemy.org/en/14/faq/sessions.html#this-session-s-transaction-has-been-rolled-back-due-to-a-previous-exception-during-flush-or-similar), recent discussion on slack's admin channel.

A commit to the database may fail for various reasons. The database transaction is rolled back; however the outer "logical" transaction that maintains the transactional state of SQLAlchemy's session won't be closed until explicitly rolled back (see referenced docs for an explanation of why that is the case). As a result, if a commit error is improperly handled, after which the session attempts to access the database, a `PendingRollbackError` will be raised. 

The proposed solution follows the recommended practice (as per SA docs) and issues a rollback in case of any error, re-raising the error for handling by the caller. However, a `PendingRollbackError` is not re-raised: I think we don't need to; instead we may want to handle specific errors that are raised due to things like constraint violation, etc. Besides, a `PendingRollbackError` will be also raised if a database disconnect occurs prior to the call to `commit` - in that case the logical transaction will be invalidated (preventing any further db access via this session); rolling it back, I think, is the right way to handle it.

If this is an acceptable approach, I'll replace all the other calls to `Session.commit` with calls to this `commit` function.

## How to test the changes?
(Select all options that apply)
- [ ] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).
- [ ] This is a refactoring of components with existing test coverage.
- [ ] Instructions for manual testing are as follows:
  1. [add testing steps and prerequisites here if you didn't write automated tests covering all your changes]

## License
- [x] I agree to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT).

### CODE REVIEW

File: lib/galaxy/model/base.py [hunk 1]
```diff
@@ -41,6 +42,25 @@
...
+        session.commit()
+    except PendingRollbackError as e:
+        err = e
+    except Exception as e:
+        err = e
+        raise
+    finally:
+        if err:
+            session.rollback()
+            log.error("Database transaction rolled back due to the following error: %s" % err)
```
Comment (nsoranzo) [inline lib/galaxy/model/base.py]:
```suggestion
            log.error("Database transaction rolled back due to the following error: %s", err)
```

**Comment by mvdbeek** [comment]
This doesn't look right to me, it is the opposite of

> Applications should ensure some system of “framing” is applied to ORM-oriented processes so that connection and transaction resources have a definitive boundary, and so that transactions can be explicitly rolled back if any failure conditions occur.

if this was the right thing to do it would surely be the default in sqlalchemy anyway.

File: lib/galaxy/model/base.py (hunk 1 above)
Comment (mvdbeek) [reply lib/galaxy/model/base.py]: If we want to do this we surely want to use log.exception ?

File: lib/galaxy/model/base.py [hunk 2]
```diff
@@ -41,6 +42,25 @@
...
+def commit(session: Union[scoped_session, Session, "SessionlessContext"]):
+    """
+    Wraps SQLAlchemy's Session.commit method. Should be called from galaxy instead of using
+    SQLAlchemy's method directly.
+    """
+    err = None
+    try:
+        session.commit()
+    except PendingRollbackError as e:
+        err = e
```
Comment (mvdbeek) [inline lib/galaxy/model/base.py:54]: So now we haven't committed and the calling context has no idea that we didn't commit. That doesn't make sense to me.

File: lib/galaxy/model/base.py [hunk 3]
```diff
@@ -41,6 +42,25 @@
...
+    except Exception as e:
+        err = e
+        raise
+    finally:
+        if err:
+            session.rollback()
```
Comment (mvdbeek) [inline lib/galaxy/model/base.py:60]: I think this belongs in the outermost layer of whatever persistent process is doing a thing, i.e the job watcher, the job queue handler, the workflow scheduler etc.

The one place where we do this correctly is the web transactions and now the celery tasks.

File: lib/galaxy/model/base.py (hunk 1 above)
Comment (jdavcs) [reply lib/galaxy/model/base.py]: Yes, I wasn't sure which one is preferable here. I'll make the change now regardless of whether we decide to go with this or not - just to make sure this suggestion is not lost.

File: lib/galaxy/model/base.py [hunk 4]
```diff
@@ -41,6 +42,25 @@
...
+            log.exception("Database transaction rolled back due to the following error: %s", err)
```
Comment (nsoranzo) [inline lib/galaxy/model/base.py]: AFAIK `log.exception` can only be used inside an `except:` branch (and it's redundant to have `err` in the log message when you use it).

Comment (jdavcs) [reply lib/galaxy/model/base.py]: You're right - it is indeed redundant to include `err`. And yes, when not used inside the `except` block, exception info is not added to the message. Back to `log.error`.

**Comment by jdavcs** [comment]
> This doesn't look right to me, it is the opposite of
> 
> > Applications should ensure some system of “framing” is applied to ORM-oriented processes so that connection and transaction resources have a definitive boundary, and so that transactions can be explicitly rolled back if any failure conditions occur.

I was referring to this as recommended practice: "it is expected that the calling code will **always** call Session.commit(), Session.rollback(), or Session.close() to correspond to the current transaction block." But yes, whether or not the proposed approach covers the "always" part is another question.

> if this was the right thing to do it would surely be the default in sqlalchemy anyway.

This may or may not be the right thing to do for *our* use case. There's [a paragraph in SA's docs](https://docs.sqlalchemy.org/en/14/faq/sessions.html#but-why-isn-t-the-one-automatic-call-to-rollback-enough-why-must-i-rollback-again) that addresses just this - why automatic rollback is not the default.

**Comment by mvdbeek** [comment]
> it is expected that the calling code will **always** call Session.commit(), Session.rollback(), or Session.close() to correspond to the current transaction block.

we can and should ensure that for our application code, but not in the commit wrapper. there's not that many entrypoints, I think we're mostly covered

File: lib/galaxy/model/base.py (hunk 2 above)
Comment (jdavcs) [reply lib/galaxy/model/base.py:54]: My reasoning was this. A `PendingRollbackError` may be raised due to the session trying to access the database (a) after a server disconnect which invalidates the outer, logical (not database) transaction; or (b) after a database transaction rollback due to failed commit, which results in the current logical transaction being in `inactive` state. 

In most cases, if a commit causes a database transaction rollback, SA will raise an error that we don't handle, which is not bad - at least we'll see the error message. But I'm sure there are places in our codebase where we have a block of code wrapped in a try/except that handles a generic `Exception` without rolling back the logical transaction, after which, as soon as we try to access the database, we'll get the `PendingRollbackError`. Here's one such example: https://github.com/galaxyproject/galaxy/blob/dev/lib/galaxy/authnz/custos_authnz.py#L361 . I'm sure it's not the only one.

By calling `commit(session)`, we at least ensure that the logical transaction is rolled back, and when the session tries to access the database after that, a new transaction begins. Of course this does not solve the issue of the silenced error raised during the commit - but that's a separate problem (that also should be eventually addressed), but it ensures that the previous error will not affect subsequent commits.

It's a partial solution only - it handles commits, but not database access in general. If a disconnect happens, or a bad commit is silenced, and then the session tries to access the database in any other way (explicit read via select or implicit read when accessing a lazy loaded attribute), we'll still have the same problem. However, given that we commit left and right (after which a new logical transaction is started as soon as the session needs one), I thought this still could be helpful as a partial solution. If you think it's not a good approach, I'm OK with closing this.

File: lib/galaxy/model/base.py (hunk 3 above)
Comment (jdavcs) [reply lib/galaxy/model/base.py:60]: Maybe. But we have multiple transactions that happen one after another (every time we call `session.commit` and then access the db again, we have a new transaction.)

**Comment by jdavcs** [comment]
> > it is expected that the calling code will **always** call Session.commit(), Session.rollback(), or Session.close() to correspond to the current transaction block.
> 
> we can and should ensure that for our application code, but not in the commit wrapper. there's not that many entrypoints, I think we're mostly covered

OK, closing this. Still, what worries me is that I don't know exactly what caused those pending rollback errors in the first place (the ones that resulted from a failed commit or flush, not invalidated after a disconnect). If it was the wrong session, then your scoping PR will have solved it. Anyway, I'll keep looking...
Thanks for reviewing!

**Pull request closed without merging**